frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java

```diff
@@ -96,18 +100,30 @@ private void rename(GitRepository gitRepository, BaseEnhancedGraphTable graphTab
       Runnable renameRunnable = () -> gitBrancher.renameBranch(branch.getName(), options.getName(),
           Collections.singletonList(gitRepository));
 
-      Runnable doInUIThreadWhenReady = new RenameBackgroundable(gitRepository,
-          branchLayout,
-          renameRunnable,
-          branch.getName(),
-          options.getName())::queue;
+      Path macheteFilePath = gitRepository.getMacheteFilePath();
 
-      // Passing same branch as a currentBranchNameIfManaged.
-      // It is a hack to prevent SlideOutBackgroundable from the branch deletion.
-      new SlideOutBackgroundable(branch,
-          gitRepository, /* currentBranchNameIfManaged */ branch, branchLayout,
-          graphTable, doInUIThreadWhenReady).queue();
+      val newBranchLayout = branchLayout.rename(branch.getName(), options.getName());
+      val branchLayoutWriter = RuntimeBinding.instantiateSoleImplementingClass(IBranchLayoutWriter.class);
 
+      graphTable.disableEnqueuingUpdates();
+      renameRunnable.run();
+
+      runWriteActionOnUIThread(() -> {
+        MacheteFileWriter.writeBranchLayout(
+            macheteFilePath,
+            branchLayoutWriter,
+            newBranchLayout,
+            /* backupOldLayout */ true,
+            /* requestor */ this);
+
+      });
+
+      // `renameRunnable` may perform some sneakily-asynchronous operations (e.g. renameBranch).
+      // The high-level method used within the runnable does not allow us to schedule the tasks after them.
+      // (Stepping deeper is not an option since we would lose some important logic or become very dependent on the internals of git4idea).
+      // Hence, we wait for the creation of the branch (with exponential backoff).
+      waitForCreationOfLocalBranch(gitRepository, options.getName());
```

Comment:
is it safe to call it here?

Reply:
Nopes, the entire `void rename` must be executed on UI thread (hence `@UIEffect`)... and this [has been caught by the CI](https://app.circleci.com/pipelines/github/VirtusLab/git-machete-intellij-plugin/8185/workflows/2540f1f9-1f08-4b7a-9a50-f0c3531d7500/jobs/8611)

Reply:
I'd move the part with `waitForCreationOfLocalBranch(...)` and `graphTable.enableEnqueuingUpdates()` to a backgroundable, probably... albeit `graphTable.enableEnqueuingUpdates()` must be put into `onFinished` (not `onSuccess`!) so that the lock is **always** released, no matter what happens in `waitForCreationOfLocalBranch`

Reply:
> I'd even remove `RenameBackgroundable` 

🙂

Reply:
Sorry for the confusion :trollface: